### DISCUSSION TITLE
Why do Forms V2 Builder and Repeater store data as JsonObject instead of JsonArray?

### DISCUSSION THREAD

**Discussion by blite** [filamentphp/filament#732] [Help]
I mainly ask because of the ability to change the order of objects  which is great.  I don't quite get how it will work that you can reload the data and the desired behavior is to retain the order in all databases when you use JsonObject.  I'm pretty sure that JsonObject is not meant to retain order even if by some accidents it does. With the choice to use JsonObect which could simply update code, I would expect an "order" attribute.

I also am slightly wondering why builders use "data" and repeaters do not.

**Comment by danharrin** [comment]
Hey,

Repeaters and Builders use keys in order to uniquely identify an item, and track its position in the item order. I've not actually seen these keys get out of order before but if you could provide an example I'd happily investigate. Using a separate `order` attribute might not be very performant as each time you change the field state, you'd need to update every order.

Builders use `data` as they also require `type` property, which is the type of block that the item is using. Repeaters only have one block type so don't require the extra property and nesting.

Hope this clears stuff up!

**Comment by blite** [comment]
What database are you using?  I am actually not seeing it working on mysql:8.0 with chrome as browser.  I suspect this could be broken by browser and database implementations as neither one has to respect the order, it does seem like the php implementation here is an anomaly in that it respects order on jsonObjects.  

In the past I have ran into bugs related to this in browsers when you used integer keys, but then not with string keys, doesn't matter here because we are using string keys, but these implementations are usually not an ordered hashmap as it is less efficient.  

I don't really feel like this is 100% a bug at this point, but I haven't yet seen the functionality actually working on a create -> load -> save again cycle.

**Reply by danharrin** [reply]
If you're having issues with it, I can recommend a fix:

```php
->afterStateHydrated(fn ($component, $set, $state) => $set($component, collect($state)
    ->mapWithKeys(fn ($item) => [Str::uuid() => $item])
    ->toArray()
))
->dehydrateStateUsing(fn ($state) => array_values($state))
```

`afterStateHydrated()` will transform a normal array into a UUID-keyed array for the lifetime of the component
`dehydrateStateUsing()` will transform the UUID-keyed array back into a normal array for saving

Maybe this should be default behaviour 🤔
